### ISSUE TITLE
Metadata: Improve XMP sidecar file support

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#2260]
**As a user with a lot of metadata in XMP sidecar files, I want PhotoPrism to read (and eventually write) more of that information, so that I can view, search, and organize my pictures based on the tags my photo software already created.**

This is an **epic** that tracks all related XMP read/write enhancements, bug reports, and pull requests. Please file specific problems and feature requests as separate issues and link them here rather than expanding this description.

Our goal is to make PhotoPrism a first-class citizen in XMP-based workflows by:

- reading a broad, well-tested set of XMP tags from both embedded and sidecar metadata;
- grouping related files via their XMP document identifier (stacking);
- importing face-region metadata written by other applications; and
- eventually writing PhotoPrism's own metadata back to XMP/EXIF.

The XMP sidecar reader was rewritten in #5563 on top of `antchfx/xmlquery` + XPath, replacing the original proof-of-concept `encoding/xml` reader and expanding sidecar tag coverage from 10 to 33 fields (title, caption, keywords/subject, GPS lat/lng/altitude, capture time with `OffsetTime` handling, camera/lens, exposure, copyright/license, document/instance IDs, and more).

### Place & Time

- [ ] #4106 — Support reading GPS information from XMP _(reader landed in #5563; values are applied on a complete rescan)_
- [ ] #2828 — Read GPS information from Apple XMP _(covered by #5563)_
- [ ] #3894 — Timestamps may be incorrect for non-standard XMP sidecars created by PhotoSync

### Labels

- [ ] #5714 — Match multi-word label names from title, caption, subject, and keyword text (links existing Labels only)
- [x] #2075 — Support for darktable's `dc:subject` _(read into the Subject field; searchable and label-generating)_
- [ ] #5710 — Import hierarchical XMP keyword tags (`lr:hierarchicalSubject`, `digiKam:TagsList`, `MicrosoftPhoto:LastKeywordXMP`) as Labels
- [ ] #1828 — Add support for digiKam pick/rating labels/tags (e.g. `digiKam:PickLabel`)
- [ ] #1143 — Automatically create labels for keywords from photo metadata (Exif/IPTC/XMP)
- [ ] #5513 — Option to disable automatic label generation from captions and titles (opt-out for the same text sources as #5714)

### Face Regions

- [ ] #554 — Import Adobe's XMP face-region metadata
- [ ] #747
- [ ] #1570 — Import Microsoft's XMP face-region metadata
- [ ] Related PR in progress: #5572 — Import MWG XMP face regions from sidecar files

### Stacking

- [ ] #4094 — Stack photos by `dc:identifier` from XMP sidecar files _(#5563 now stores the identifier as the photo `UUID`; grouping separate files by a non-canonical identifier from a sidecar is still open)_

### Writing & Embedding

- [ ] #402 — Write PhotoPrism metadata to XMP sidecars or EXIF
- [ ] #243 — Embed XMP metadata in JPEG files

### Completed

- [x] #5563 — Index: Improve support for metadata in XMP sidecar files (reader rewrite, 33-field coverage)
- [x] #1161 — Add support for XMP sidecar `CreateDate` and `Keywords`
- [x] #1873 — Add support for reading the `fstop:favorite` flag
- [x] #636 — Read metadata from Apple iPhoto / Photos app XMP

### Documentation

- Developer Guide › Metadata › XMP: https://docs.photoprism.app/developer-guide/metadata/xmp/
- Reader implementation & supported-tag coverage: [`internal/meta/README.md`](https://github.com/photoprism/photoprism/blob/develop/internal/meta/README.md)
- XMP specification (ISO 16684-1) and overview: https://www.adobe.com/products/xmp.html

**Comment by jmalm** [comment]
Have you considered using ExifTool in some way for reading / writing image metadata (both from the images and to/from XMP sidecar files)? For example using https://github.com/barasher/go-exiftool? (go-exiftool has a GPL 3 license, which might not be compatible with the licensing strategy of Photoprism. ExifTool, on the other hand, looks like it has a very permissive license.)

My impression is that ExifTool has become something of a standard reference tool when it comes to image metadata, so it may help even more than yq.

For reference, Librephotos does it this way. (I implemented writing to XMP sidecar files and made some smaller changes to the reading as well in that project, but the choice of ExifTool was made earlier.)

**Comment by lastzero** [comment]
We already use Exiftool, but it's an external Perl script and XML isn't hard to parse as such. It's just that the built-in XML support in Go is pretty bad, at least last time I checked. See my notes in the Developer Guide.

**Comment by jmalm** [comment]
(The following is some testing and reasoning...)

I guess one of the nice things with using ExifTool is that you don't have to care as much about the structure of the tags? I.e. you can read rating by `exiftool -Rating FILENAME` and write it by `exiftool -Rating=4 FILENAME` in an XMP sidecar like the following:

```xml
<x:xmpmeta xmlns:x='adobe:ns:meta/' x:xmptk='Image::ExifTool 10.10'>
<rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'>

 <rdf:Description rdf:about=''
  xmlns:dc='http://purl.org/dc/elements/1.1/'>
  <dc:creator>
   <rdf:Seq>
    <rdf:li>Jakob Malm</rdf:li>
   </rdf:Seq>
  </dc:creator>
  <dc:rights>
   <rdf:Alt>
    <rdf:li xml:lang='x-default'>Jakob Malm</rdf:li>
   </rdf:Alt>
  </dc:rights>
 </rdf:Description>

 <rdf:Description rdf:about=''
  xmlns:xmp='http://ns.adobe.com/xap/1.0/'>
  <xmp:Rating>4</xmp:Rating>
 </rdf:Description>
</rdf:RDF>
</x:xmpmeta>
```

With yq command line, reading would be something like `yq -p=xml ".xmpmeta.RDF.Description[] | select(.Rating) | .Rating" FILENAME`.
Writing could be accomplished by `yq -p=xml -o=xml "(.xmpmeta.RDF.Description[] | select(.Rating) | .Rating) = 4" FILENAME`, but it seems the entire file is rewritten (I guess that is probably the case with exiftool too), **without namespaces**:
```xml
<xmpmeta x="adobe:ns:meta/" xmptk="Image::ExifTool 12.16">
  <RDF rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
    <Description about="" dc="http://purl.org/dc/elements/1.1/">
      <creator>
        <Seq>
          <li>Jakob Malm</li>
        </Seq>
      </creator>
      <rights>
        <Alt>
          <li lang="x-default">Jakob Malm</li>
        </Alt>
      </rights>
    </Description>
    <Description about="" xmp="http://ns.adobe.com/xap/1.0/">
      <Rating>4</Rating>
    </Description>
  </RDF>
</xmpmeta>
```

I haven't found a way to not have to specify (and know!) the "absolute" path to the tag. This may be ok and perhaps even desired, but I think the missing namespaces might be a problem.

On the _other_ hand, two nice things with using `yq` would be
- being able to compile the functionality into the application (via `yqlib`)
- being able to use the same library for XMP, YAML and JSON sidecar and metadata files.

**Comment by lastzero** [comment]
An alternative library to take a look at is https://github.com/beevik/etree

XMP-related issues that may depend on this:
- #554
- #1570
- #1873
- #2828

**Comment by lastzero** [comment]
I've turned this into an epic to keep track of related follow-up enhancements and feature requests. Support for improved metadata indexing from XMP sidecar files has been merged into `develop` and will ship with our next stable release.
